app/src/main/java/org/oppia/android/app/home/HomeViewModel.kt

```diff
@@ -0,0 +1,164 @@
+import org.oppia.android.util.parser.TopicHtmlParserEntityType
+import org.oppia.android.util.system.OppiaClock
+import javax.inject.Inject
+
+private const val PROFILE_AND_ONGOING_STORY_COMBINED_PROVIDER_ID = "profile+ongoingStoryList"
+private const val HOME_FRAGMENT_COMBINED_PROVIDER_ID = "profile+ongoingStoryList+topicListProvider"
+
+/** [ViewModel] for layouts in home fragment . */
+@FragmentScope
+class HomeViewModel @Inject constructor(
+  private val activity: AppCompatActivity,
+  private val fragment: Fragment,
+  private val oppiaClock: OppiaClock,
+  private val logger: ConsoleLogger,
+  private val internalProfileId: Int,
+  private val intentFactoryShim: IntentFactoryShim,
+  private val profileManagementController: ProfileManagementController,
+  private val topicListController: TopicListController,
+  @TopicHtmlParserEntityType private val topicEntityType: String,
+  @StoryHtmlParserEntityType private val storyEntityType: String
+) : ObservableViewModel() {
+
+  private val profileId: ProfileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+  private val limit = activity.resources.getInteger(R.integer.promoted_story_list_limit)
+
+  private val profileDataProvider: DataProvider<Profile> by lazy {
+    profileManagementController.getProfile(profileId)
+  }
+
+  private val ongoingStoryListSummaryDataProvider: DataProvider<OngoingStoryList> by lazy {
+    topicListController.getOngoingStoryList(profileId)
+  }
+
+  private val topicListSummaryDataProvider: DataProvider<TopicList> by lazy {
+    // TODO: once #2253 is merged change this function call
+    topicListController.getTopicList2()
+  }
+
+  private val homeItemViewModelListDataProvider: DataProvider<List<HomeItemViewModel>> by lazy {
+    // This will block until all data providers return initial results (which may be default
+    // instances). If any of the data providers are pending or failed, the combined result will also
+    // be pending or failed.
+    profileDataProvider.combineWith(
+      ongoingStoryListSummaryDataProvider,
+      PROFILE_AND_ONGOING_STORY_COMBINED_PROVIDER_ID
+    ) { profile, ongoingStoryList ->
+      listOfNotNull(
+        computeWelcomeViewModel(profile),
+        computePromotedStoryListViewModel(ongoingStoryList)
+      )
+    }.combineWith(
+      topicListSummaryDataProvider,
+      HOME_FRAGMENT_COMBINED_PROVIDER_ID
+    ) { homeItemViewModelList, topicList ->
+      homeItemViewModelList + listOf(AllTopicsViewModel()) + computeTopicSummaryItemViewModelList(
+        topicList
+      )
+    }
+  }
+
+  // Resulting LiveData to bind to the outer RecyclerView & that contains all ViewModels.
+  val homeItemViewModelListLiveData: LiveData<List<HomeItemViewModel>> by lazy {
+    Transformations.map(homeItemViewModelListDataProvider.toLiveData()) { itemListResult ->
+      if (itemListResult.isFailure()) {
+        logger.e(
+          "HomeFragment",
+          "Failed to retrieve fragment",
```

Comment:
I agree this is vague, but I'm unsure of how to get more information; this one was modeled off of an error in the previous implementation of HomeFragmentPresenter. This would fail if any of the child ViewModels (welcome, promoted story list, or any of the topic summaries) fails -- are we able to access the inner view models if itemsListResult fails (I suspect not)?

My other thought was to check each DataProvider to look for the error but I don't think that's an accurate explanation for this specific Result failing since the fetching may be successful / completed by then.

Reply:
I think it's more an issue with the actual string--there isn't a fragment being retrieved to fail. Can you make that more clear?

Reply:
Doesn't this occur though when the data is not retrieved or (pending or error)? Is this better?

Reply:
This is clearer now--thanks!